# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/generic/nsHTMLCanvasFrame.cpp
# Commit: 5be8557c4721
# Full Hash: 5be8557c4721b4d1a2ac07bf5ae0516d9f2da2bc
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2021-12-10 09:10:34
# Regressor Bug: 1736177
# File Overlap Count: 2
# Description:
#   Bug 1736177 - Part 11. Plumb in OffscreenCanvasDisplayHelper. r=jgilbert
#   
#   This patch integrates OffscreenCanvasDisplayHelper with
#   HTMLCanvasElement, OffscreenCanvas and nsDisplayCanvas to allow
#   asynchronous display of an OffscreenCanvas.
# ==============================================================================

diff -r 49f8b4205a46 -r 5be8557c4721 layout/generic/nsHTMLCanvasFrame.cpp
--- a/layout/generic/nsHTMLCanvasFrame.cpp	Thu Dec 09 19:25:28 2021 +0000
+++ b/layout/generic/nsHTMLCanvasFrame.cpp	Thu Dec 09 19:25:28 2021 +0000
@@ -108,6 +108,31 @@
         static_cast<HTMLCanvasElement*>(mFrame->GetContent());
     element->HandlePrintCallback(mFrame->PresContext());
 
+    if (element->IsOffscreen()) {
+      // If we are offscreen, then we display via an ImageContainer which is
+      // updated asynchronously, likely from a worker thread. There is nothing
+      // to paint until the owner attaches it.
+      RefPtr<ImageContainer> container = element->GetImageContainer();
+      if (container) {
+        nsHTMLCanvasFrame* canvasFrame =
+            static_cast<nsHTMLCanvasFrame*>(mFrame);
+        nsIntSize canvasSizeInPx = canvasFrame->GetCanvasSize();
+        IntrinsicSize intrinsicSize =
+            IntrinsicSizeFromCanvasSize(canvasSizeInPx);
+        AspectRatio intrinsicRatio =
+            IntrinsicRatioFromCanvasSize(canvasSizeInPx);
+        nsRect area =
+            mFrame->GetContentRectRelativeToSelf() + ToReferenceFrame();
+        nsRect dest = nsLayoutUtils::ComputeObjectDestRect(
+            area, intrinsicSize, intrinsicRatio, mFrame->StylePosition());
+        LayoutDeviceRect bounds = LayoutDeviceRect::FromAppUnits(
+            dest, mFrame->PresContext()->AppUnitsPerDevPixel());
+        aManager->CommandBuilder().PushImage(this, container, aBuilder,
+                                             aResources, aSc, bounds, bounds);
+      }
+      return true;
+    }
+
     switch (element->GetCurrentContextType()) {
       case CanvasContextType::Canvas2D:
       case CanvasContextType::WebGL1:
@@ -328,6 +353,10 @@
       return;
     }
 
+    if (canvas->IsOffscreen()) {
+      return;
+    }
+
     RefPtr<CanvasRenderer> renderer = new CanvasRenderer();
     if (!canvas->InitializeCanvasRenderer(aBuilder, renderer)) {
       return;
